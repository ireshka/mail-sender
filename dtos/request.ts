import {
  IsString,
  IsEmail,
  Min,
  Max,
  IsInt,
  IsArray,
  ValidateNested,
} from "class-validator";
import { ISubmitRequest, IVote } from "../types/Survey";

export class Vote implements IVote {
  @IsString()
  id!: string;

  @IsInt()
  @Min(1)
  @Max(5)
  vote!: number;
}

export class SubmitRequest implements ISubmitRequest {
  @IsString()
  @IsEmail()
  email!: string;

  @IsArray()
  @ValidateNested({ each: true })
  votes!: Vote[];
}
