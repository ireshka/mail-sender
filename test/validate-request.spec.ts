import { validate } from "class-validator";
import { Vote, SubmitRequest } from "../dtos/request";

function generateRequest(overrides: { email?: any; votes?: any }) {
  const request = new SubmitRequest();
  request.email =
    overrides.email === undefined ? "mihau@gmail.com" : overrides.email;
  request.votes = overrides.votes === undefined ? [] : overrides.votes;

  return request;
}

test("passes with correct request", async () => {
  const request = generateRequest({});
  expect(await validate(request)).toEqual([]);
});

test("fails without email", async () => {
  const request = generateRequest({ email: null });
  expect(await validate(request)).toEqual([
    expect.objectContaining({
      constraints: {
        isEmail: "email must be an email",
        isString: "email must be a string",
      },
      property: "email",
    }),
  ]);
});

test("fails with incorrect email", async () => {
  const request = generateRequest({ email: "mihaugmail.com" });
  expect(await validate(request)).toEqual([
    expect.objectContaining({
      constraints: {
        isEmail: "email must be an email",
      },
      property: "email",
    }),
  ]);
});

test("fails with empty votes", async () => {
  const request = generateRequest({ votes: null });
  expect(await validate(request)).toEqual([
    expect.objectContaining({
      children: [
        expect.objectContaining({
          constraints: {
            nestedValidation:
              "each value in nested property votes must be either object or array",
          },
          property: "votes",
        }),
      ],
    }),
  ]);
});

test("fails with incorrect answer", async () => {
  const request = generateRequest({ votes: [new Vote()] });
  expect(await validate(request)).toMatchSnapshot();
});
