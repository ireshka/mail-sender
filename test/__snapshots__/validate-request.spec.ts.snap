// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`fails with incorrect answer 1`] = `
Array [
  ValidationError {
    "children": Array [
      ValidationError {
        "children": Array [
          ValidationError {
            "children": Array [],
            "constraints": Object {
              "isString": "id must be a string",
            },
            "property": "id",
            "target": Vote {},
            "value": undefined,
          },
          ValidationError {
            "children": Array [],
            "constraints": Object {
              "isInt": "vote must be an integer number",
              "max": "vote must not be greater than 5",
              "min": "vote must not be less than 1",
            },
            "property": "vote",
            "target": Vote {},
            "value": undefined,
          },
        ],
        "property": "0",
        "target": Array [
          Vote {},
        ],
        "value": Vote {},
      },
    ],
    "property": "votes",
    "target": SubmitRequest {
      "email": "mihau@gmail.com",
      "votes": Array [
        Vote {},
      ],
    },
    "value": Array [
      Vote {},
    ],
  },
]
`;
